OpenRA.Mods.Common/Traits/Attack/AttackBase.cs

```diff
@@ -217,6 +214,19 @@ public bool HasAnyValidWeapons(Target t)
 			return false;
 		}
 
+		public bool HasAnyValidWeaponTargetingCenter(Target t)
+		{
+			if (!HasAnyValidWeapons(t))
+				return false;
+
+			// PERF: Avoid LINQ.
+			foreach (var armament in Armaments)
+				if (armament.AttackTargetCenter && armament.Weapon.IsValidAgainst(t, self.World, self))
+					return true;
+
+			return false;
+		}
```

Comment:
It could be better to add targetingCenter argument to HasAnyValidWeapons. Now Armaments loop is done twice. Otherwise it should be fixed - it can return True for OutOfAmmo Armament. HasAnyValidWeapons can return true for !AttackTargetCenter Armament, here we could select another which is OutOfAmmo.